### ISSUE TITLE
Releasing SymPy 1.9

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#22008]
This issue is to track the release of SymPy 1.9. The release issue for 1.8 was #21264

Release notes for 1.9 are here:
https://github.com/sympy/sympy/wiki/Release-Notes-for-1.9

- [x] Make sure latest versions of Python and other dependencies are tested in CI
- [x] Check authors:
  - [x] run bin/mailmap_update.py
  - [x] run bin/authors_update.py
  - [x] Open a PR to update .mailmap and AUTHORS
  - [x] Email all unrecognised email addresses and tag unknown authors in GH
  - [x] Merge to master before making a release branch.
- [x] Open a PR for master:
  - [x] Change the version in sympy/release.py on master from 1.9.dev to 1.10.dev
  - [x] Update previous versions for branchmarks job in runtests.yml and asv.conf.actions.json from 1.8 to 1.9
- [x] Create a 1.9 branch:
  - [x] Change version in sympy/release.py on 1.9 branch from 1.9.dev to 1.9rc1
  - [x] Change the release version in release.yml to 1.9rc1
  - [x] Disable Travis benchmarks on 1.9
  - [x] Create the release notes page for 1.10
- [x] Push the 1.9 branch 
- [x] Merge master version PR after the 1.9 branch is pushed
- [x] Email mailing list about release and any release blockers
- [x] Port release actions workflow from 1.8
- [x] Update versions in release actions workflow
- [ ] Port release actions workflow to master
- [ ] Edit release notes
  - [ ] Note deprecations and compatibility breaks
  - [ ] Add highlights of the release
  - [ ] Make formatting consistent
  - [ ] Improve badly worded release notes
- [ ] Resolve release blocker issues (on both master and 1.9 branch)
- [ ] Check blockers from old releases

**Comment by oscargus** [comment]
Is there a process to determine which PR:s one should try to get in before the release? That could probably help at least me both prioritizing and to some extent get motivation to finish them.

(Another, slightly related questions, have you considered having some sort of focused activity to clean up issues and handle PRs? There is a quite big list and some focused maintenance effort could probably make some difference. I realize that this is a community effort, but I do not really find that I am in a position to take the lead here. Maybe one should announce a weekend or something as cleaning up weekend and then one can split issues and PRs between us and see what can be merged/closed/updated.)

**Comment by oscarbenjamin** [comment]
> Is there a process to determine which PR:s one should try to get in before the release?

I'm not sure there is a documented process but basically it is this:

1. If at any time an issue is identified as a regression since a previous release (especially if it is a regression on master relative to the previous release) then it should have the regression label and should be given the milestone for the next release:

    https://github.com/sympy/sympy/issues?q=is%3Aopen+is%3Aissue+label%3Aregression
    https://github.com/sympy/sympy/milestone/59

2. If some work from between two releases is somewhat unfinished or unresolved and you think it should get sorted before the release then relevant issues and PRs should also be added to the milestone.

3. Before releasing we should make sure that all issues/PRs with the release milestone are resolved in one way or another.

4. Once a release branch is made only regressions should really be added to the milestone or merged into the release branch (although there can be exceptions).

5. If you are unsure whether the milestone should be applied then we can discuss here in this issue.

I've already made a release branch for 1.9 now. If you know of any issues to resolve then add the milestone or discuss them here. To be fixed in the release now though there needs to be separate PRs for master and for the release branch.

> Another, slightly related questions, have you considered having some sort of focused activity to clean up issues and handle PRs?

That's an important question but I think it should have its own issue or should perhaps be discussed on the mailing list rather than GitHub.

**Comment by oscargus** [comment]
OK! I add milestones to what I think are relevant issues. Feel free to remove them...

Currently merges go to 1.10, right? Is the idea to explicitly merge them to 1.9 separately? Is there a simple way to do that (considering release notes etc)?

**Comment by oscarbenjamin** [comment]
> Currently merges go to 1.10, right?

Yes, current master is for 1.10

> Is the idea to explicitly merge them to 1.9 separately? Is there a simple way to do that (considering release notes etc)?

The way to do it is to first open a PR for master and if that's good to merge then open a new PR rebased onto the 1.9 branch. The release note should go on the PR for the 1.9 branch if it is going to be merge into 1.9. Only I should merge PRs on the 1.9 branch.

**Comment by asmeurer** [comment]
> (Another, slightly related questions, have you considered having some sort of focused activity to clean up issues and handle PRs? There is a quite big list and some focused maintenance effort could probably make some difference. I realize that this is a community effort, but I do not really find that I am in a position to take the lead here. Maybe one should announce a weekend or something as cleaning up weekend and then one can split issues and PRs between us and see what can be merged/closed/updated.)

Honestly, I don't think there is anyone who can look at an arbitrary issue and tell if it can be closed or not. There are too many different parts of SymPy and no one is an expert in all of them. So rather than thinking about it like that, it's better to think about looking at the subset of issues relating to the submodules you are familiar with. There are a lot of open issues, so even clearing out a subset would be significant and helpful.

**Comment by oscargus** [comment]
> Honestly, I don't think there is anyone who can look at an arbitrary issue and tell if it can be closed or not.

I see your point. However, I've had some success just looking at old issues (and PRs), running the example and see if it still is a bug or not, if it works: add a test. At the moment that becomes a quite random exercise though. Clearly, it makes sense to do that on some sort of submodule basis.

Similarly with PRs, there are quite a lot of PRs that are "almost ready". Some are not merged for no apparent reason (apart from people being busy and then forgetting about it), some may have minor merge conflicts, some just needs to add tests. 

By having a focused community effort I believe that there can be a reduction in issues and PRs which would be beneficial. (I've merged and/or closed a handful of PRs today, so now there are less than 500.) Just having some way to mark them would be good as well, basically, "here is a PR that looks good and there is only a minimum amount of work left, lets try to get it is". Similarly with some issues where the solutions is basically there. Now, this often leads to someone picking it up, but not always.

Basically, I'm just thinking that it is a bit unfortunate that there is quite a lot of good/reasonable work that has been made and it gets stuck in the merge process. I understand why, but a directed effort could make a significant difference. Especially now that 1.9 is (almost) released and there is time to sort out any negative side effects from merging things that were not obvious.

(With that said, it is not like I have unlimited time either, but it would feel a bit better spending the time if one knew that it was part of a bigger effort.)

**Comment by moorepants** [comment]
I think if there was a leader that started the PR/issue merging/closing effort, defined what it was, made it easy for people to join in, then it could happen. This is more a herding people type of issue to solve than anything technical. There are all kinds of methods for getting a group of volunteers focused on a collective task. You could start by organizing quarterly sprints or something.

**Comment by oscargus** [comment]
Definitely. I was primarily probing for interest among the most involved persons, I believe that some of the core developers should be in, as well as people having a good knowledge of particular modules.

We'll see, but so far there's been no "Oh, that is a great idea!", and maybe that is the current state. (I shouldn't complain, I've been inactive for a couple of years and basically slipped back in when there was some bug I had probably contributed to). But I may at least drop an email to the list if I get around to it. Hacktober coming up if nothing else...

**Comment by oscarbenjamin** [comment]
I think that a drive to better manage issues and PRs is a great idea. Just as I said above it's probably best discussed on the mailing list and not on this issue which I created to keep track of various tasks involved in the release process  :)

I suggest to discuss on the mailing list and then maybe set up a video call for people who are interested (I would be).
